### DISCUSSION TITLE
How can I construct a `Query` value? (not a `QueryState` value)

### DISCUSSION THREAD

**Discussion by bbarker** [bevyengine/bevy#10430] [Q&A] [answered]
`QueryState` values appear to be stateful, so I don't really want to pass those around, especially when the intention is to re-use them. But I'm working in a context (serialization macros) where I need to construct Queries. [world.query](https://docs.rs/bevy/latest/bevy/ecs/world/struct.World.html#method.query) returns a `QueryState`. Is there a way to construct a `Query` instead?

**Comment by tim-blackbird** [comment, marked as answer]
I'm not sure why `QueryState` being 'stateful' matters here.
Keep in mind that `Query` is just a wrapper around `QueryState`.
You can construct a `Query` via [`SystemState`](https://docs.rs/bevy/latest/bevy/ecs/system/struct.SystemState.html).

**Reply by bbarker** [reply]
Thanks, I'll take a look at `SystemState`. To clarify a bit about my apprehension, I had a code snippet that looks like this, where `self` is &`QueryState`:

```rust
        let count = self.iter().count();
        match serializer.serialize_seq(Some(count)) {
            Ok(mut serseq) => {
                self.iter().for_each(|ent_comp| {
                    serseq.serialize_element(&ent_comp).unwrap();
                });
                serseq.end()
            }
            Err(er) => Err(er),
        }
    }
 ```
 
Since `QueryState` needs to be mutable to iterate over it, it just seemed like the interface might be hiding some mutation that occurred and I might not be able to reuse it.

**Reply by bbarker** [reply]
Well, I see that iter() is the same for Query in regard to needing mutable access. So maybe I should just stick with QueryState and explore along those lines.
